library/src/main/java/kwasm/format/text/FunctionType.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kwasm.format.text
+
+import kwasm.ast.FunctionType
+import kwasm.ast.Param
+import kwasm.format.ParseException
+import kwasm.format.text.token.Keyword
+import kwasm.format.text.token.Paren
+import kwasm.format.text.token.Token
+
+fun List<Token>.parseFunctionType(currentIndex: Int): ParseResult<FunctionType> {
+    var parsedTokens = 0
+    val openParen = this[currentIndex]
+    if (openParen !is Paren.Open) {
+        throw ParseException("Invalid FunctionType: Expecting ( token", openParen.context)
+    }
+    parsedTokens++
+    val keyword = this[currentIndex + 1]
+    if (keyword !is Keyword || keyword.value != "func") {
+        throw ParseException("Invalid FunctionType: Expecting func token", keyword.context)
+    }
+    parsedTokens++
+    var nextParamIndex = currentIndex + 2
+    val paramList = mutableListOf<Param>()
+    while (this[nextParamIndex] is Paren.Open) {
+        val paramKeyword = this[nextParamIndex + 1]
+        if (paramKeyword is Keyword && paramKeyword.value == "param") {
+            val parsedParamResult = this.parseParam(nextParamIndex)
+            paramList.add(parsedParamResult.astNode)
+            nextParamIndex += parsedParamResult.parseLength
+            parsedTokens += parsedParamResult.parseLength
+        } else {
+            break
+        }
+    }
```

Comment:
It might be worth extracting this out to its own `List<Token>.parseParamList(..): ParseResult<List<Param>>` extension function.  It'd make the parser for FunctionType that much cleaner (and parsing a param list can be tested independently.

Reply:
Done